R2API.Rules/RuleCatalogExtras.cs

```diff
@@ -0,0 +1,345 @@
+        }
+
+        static void AddRuleInternal(RuleDef ruleDef, int ruleCategoryDefIndex)
+        {
+            if (!categoryToCustomRules.ContainsKey(ruleCategoryDefIndex))
+            {
+                categoryToCustomRules[ruleCategoryDefIndex] = new List<RuleDef>();
+            }
+            categoryToCustomRules[ruleCategoryDefIndex].Add(ruleDef);
+        }
+    }
+
+    /// <summary>
+    /// Finds a custom ruleDef by comparing the ruleDef's global name with <paramref name="ruleDefGlobalName"/>
+    /// </summary>
+    /// <param name="ruleDefGlobalName">The global name of the ruleDef to find</param>
+    /// <returns>The rule def, null if it wasnt found</returns>
+    public static RuleDef FindCustomRuleDef(string ruleDefGlobalName)
+    {
+        SetHooks();
+
+        if(RuleCatalog.availability.available)
+        {
+            return RuleCatalog.FindRuleDef(ruleDefGlobalName);
+        }
+
+        return FindRuleDefInternal(ruleDefGlobalName);
+
+        static RuleDef FindRuleDefInternal(string ruleDefGlobalName)
+        {
+            foreach (RuleDef ruleDef in categoryToCustomRules.Values.SelectMany(x => x))
+            {
+                if (string.Equals(ruleDefGlobalName, ruleDef.globalName, StringComparison.OrdinalIgnoreCase))
+                {
+                    return ruleDef;
+                }
+            }
+            return null;
+        }
+    }
+
+    /// <summary>
+    /// Finds a custom RuleChoiceDef by comparing the ruleChoiceDef's global name with <paramref name="ruleChoiceDefGlobalName"/>
+    /// </summary>
+    /// <param name="ruleChoiceDefGlobalName">The global name of the rule choice def to find</param>
+    /// <returns>The rule choice def, null if it wasnt found</returns>
+    public static RuleChoiceDef FindCustomChoiceDef(string ruleChoiceDefGlobalName)
+    {
+        SetHooks();
+
+        if(RuleCatalog.availability.available)
+        {
+            return RuleCatalog.FindChoiceDef(ruleChoiceDefGlobalName);
+        }
+
+        return FindChoiceDefInternal(ruleChoiceDefGlobalName);
+
+        static RuleChoiceDef FindChoiceDefInternal(string ruleChoiceDefGlobalName)
+        {
+            foreach (RuleChoiceDef choiceDef in categoryToCustomRules.Values.SelectMany(x => x).SelectMany(x => x.choices))
+            {
+                if (string.Equals(ruleChoiceDefGlobalName, choiceDef.globalName, StringComparison.OrdinalIgnoreCase))
+                {
+                    return choiceDef;
+                }
+            }
+            return null;
+        }
+    }
+
+    #region hooks
+    internal static void SetHooks()
+    {
+        if(_hooksEnabled)
+        {
+            return;
+        }
+
+        //Due to the way how ResourceAvailability's onAvailable subscription works, we dont want to subscribe if the catalog has alreaddy finished initializing, otherwise it'll call immediatly and the entries will be added again.
```

Comment:
Isnt that wrong? the subscription doesnt seem to trigger the event to be raised

Looks like CallWhenAvailable should be used instead since this one is public and onAvailable isnt

Reply:
Seems to be a result of publicizing, going to fix it now